### ISSUE TITLE
Support python 3.12; allow up-to-date versions of scipy

### ISSUE THREAD

**Issue by kalekundert** [pymanopt/pymanopt#269]
Trying to install pymanopt in a python 3.12 environment results in the following error:
```console
$ python --version
Python 3.12.2
$ python -m pip install pymanopt
Collecting pymanopt
  Using cached pymanopt-2.2.0-py3-none-any.whl.metadata (7.0 kB)
Requirement already satisfied: numpy>=1.16.0 in ./.pyenv/versions/3.12.2/lib/python3.12/site-packages (from pymanopt) (1.26.4)
Collecting scipy<1.10,>=1.0 (from pymanopt)
  Using cached scipy-1.9.3.tar.gz (42.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      + meson setup /tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d /tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d/.mesonpy-0j6kpvgq -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d/.mesonpy-0j6kpvgq/meson-python-native-file.ini
      The Meson build system
      Version: 1.4.0
      Source dir: /tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d
      Build dir: /tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d/.mesonpy-0j6kpvgq
      Build type: native build
      Project name: SciPy
      Project version: 1.9.3
      C compiler for the host machine: /usr/bin/clang (clang 17.0.6 "clang version 17.0.6")
      C linker for the host machine: /usr/bin/clang ld.bfd 2.42.0
      C++ compiler for the host machine: /usr/bin/clang++ (clang 17.0.6 "clang version 17.0.6")
      C++ linker for the host machine: /usr/bin/clang++ ld.bfd 2.42.0
      Host machine cpu family: x86_64
      Host machine cpu: x86_64
      Compiler for C supports arguments -Wno-unused-but-set-variable: YES
      Compiler for C supports arguments -Wno-unused-but-set-variable: YES (cached)
      Compiler for C supports arguments -Wno-unused-function: YES
      Compiler for C supports arguments -Wno-conversion: YES
      Compiler for C supports arguments -Wno-misleading-indentation: YES
      Compiler for C supports arguments -Wno-incompatible-pointer-types: YES
      Library m found: YES
      Fortran compiler for the host machine: gfortran (gcc 14.1.1 "GNU Fortran (GCC) 14.1.1 20240507")
      Fortran linker for the host machine: gfortran ld.bfd 2.42.0
      Compiler for Fortran supports arguments -Wno-conversion: YES
      Program cython found: YES (/tmp/pip-build-env-4vx18uvk/overlay/bin/cython)
      Program pythran found: YES (/tmp/pip-build-env-4vx18uvk/overlay/bin/pythran)
      Program cp found: YES (/usr/bin/cp)
      Program python found: YES (/home/kale/.pyenv/versions/3.12.2/bin/python)

      ../meson.build:85:2: ERROR: Problem encountered: Your Python version is too new. SciPy 1.9 supports Python 3.8-3.11; if you are trying to build from source for the most recent SciPy version you may hit this error as well. Please build from the `main` branch on GitHub instead.

      A full log can be found at /tmp/pip-install-xfmidx2g/scipy_d7e96251adee470bb3b850027a86652d/.mesonpy-0j6kpvgq/meson-logs/meson-log.txt
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
As the error message suggests, the proximate cause of the error is the `"scipy>=1.0,<1.10"` version requirement.  I presume that scipy 1.10 caused some sort of problem, but is there any way to either (i) see if later versions of scipy solve that problem or (ii) work around it?  Right now this is just an inconvenience (I'll use python 3.11 instead of 3.12), but over time this could make it impossible to use pymanopt in modern python environments.

---

On a related note, if there's interest, I could submit a PR that would add `3.x` to the GitHub Actions test matrix.  This would then automatically run the tests on the most recent version of python.

**Comment by kburns** [comment]
+1 for this if anyone knows how to approach it!

**Comment by kalekundert** [comment]
This issue should be fixed by #270.  The fix hasn't made it into a PyPI release yet, but you can get it by installing directly from github: `pip install git+https://github.com/pymanopt/pymanopt.git`

**Comment by NicolasBoumal** [comment]
This should be fixed with release 2.2.1, now reflected on pypi.
